File: Example/Tests/MediaControlTests.m
//
//  MediaControlTests.m
//  Clappr
//
//  Created by Gustavo Barbosa on 12/18/14.
//  Copyright (c) 2014 globo.com. All rights reserved.
//

#import <Clappr/Clappr.h>
#import <objc/runtime.h>

SPEC_BEGIN(MediaControl)

describe(@"Media Control", ^{

    __block CLPMediaControl *mediaControl;
    __block CLPContainer *container;
    __block CLPPlayback *playback;

    NSURL *sourceURL = [NSURL URLWithString:@"http://globo.com/video.mp4"];

    beforeEach(^{
        playback = [[CLPPlayback alloc] initWithURL:sourceURL];
        container = [[CLPContainer alloc] initWithPlayback:playback];
        mediaControl = [[CLPMediaControl alloc] initWithContainer:container];
    });

    describe(@"Play", ^{

        it(@"should contain a play/pause button embed in its view", ^{
            UIButton *playPauseButton = mediaControl.playPauseButton;
            [[playPauseButton.superview should] equal:mediaControl.view];
        });

        it(@"should be triggered after touch the button when it is not playing", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(NO)];

            [[mediaControl should] receive:@selector(play)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should call container play after its play method has been called", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(NO)];

            [[container should] receive:@selector(play)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should trigger playing event after listen to container's play event", ^{

            __block BOOL callbackWasCalled = NO;
            [mediaControl once:CLPMediaControlEventPlaying callback:^(NSDictionary *userInfo) {
                callbackWasCalled = YES;
            }];

            [container trigger:CLPContainerEventPlay];

            [[theValue(callbackWasCalled) should] beTrue];
        });
    });

    describe(@"Pause", ^{

        it(@"should be triggered after touch the button when it is playing", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(YES)];

            [[mediaControl should] receive:@selector(pause)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should call container pause after its pause method has been called", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(YES)];

            [[container should] receive:@selector(pause)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should trigger 'not playing event' after listen to container's pause event", ^{

            __block BOOL callbackWasCalled = NO;
            [mediaControl once:CLPMediaControlEventNotPlaying callback:^(NSDictionary *userInfo) {
                callbackWasCalled = YES;
            }];

            [container trigger:CLPContainerEventPause];

            [[theValue(callbackWasCalled) should] beTrue];
        });

    });

    describe(@"Current Time", ^{

        it(@"should contain a label displaying the current playback time", ^{
            UILabel *currentTimeLabel = [mediaControl valueForKey:@"_currentTimeLabel"];
            [[currentTimeLabel.text should] equal:@"0:00"];
        });

        pending(@"should update its value after listen to playback's current time update", ^{

        });

    });
});

SPEC_END

File: Example/Tests/MediaControlTests.swift
import Quick
import Nimble
import Clappr

class MediaControlTests: QuickSpec {
    
    override func spec() {
        describe("MediaControl") {
            let sourceUrl = NSURL(string: "http://globo.com/video.mp4")!
            var container: Container!
            var playback: StubedPlayback!
            
            beforeEach() {
                playback = StubedPlayback(url: sourceUrl)
                container = Container(playback: playback)
            }
            
            context("Initialization") {
                
                it("Should have a init method to setup with container") {
                    let mediaControl = MediaControl.initFromNib()
                    mediaControl.setup(container)
                    
                    expect(mediaControl).toNot(beNil())
                    expect(mediaControl.container) == container
                }
            }
            
            context("Behavior") {
                var mediaControl: MediaControl!
                
                beforeEach() {
                    mediaControl = MediaControl.initFromNib()
                    mediaControl.setup(container)
                }
                
                context("Visibility") {
                    it("Should start with controls hidden") {
                        expect(mediaControl.controlsOverlayView.alpha) == 0
                        expect(mediaControl.controlsWrapperView.alpha) == 0
                        expect(mediaControl.controlsHidden).to(beTrue())
                    }
                    
                    it("Should show it's control after when media control is enabled on container") {
                        container.mediaControlEnabled = true
                        
                        expect(mediaControl.controlsOverlayView.alpha) == 1
                        expect(mediaControl.controlsWrapperView.alpha) == 1
                        expect(mediaControl.controlsHidden).to(beFalse())
                    }
                    
                    it("Should hide it's control after hide is called and media control is enabled") {
                        container.mediaControlEnabled = true
                        mediaControl.hide()
                        
                        expect(mediaControl.controlsOverlayView.alpha) == 0
                        expect(mediaControl.controlsWrapperView.alpha) == 0
                        expect(mediaControl.controlsHidden).to(beTrue())
                    }
                    
                    it("Should show it's control after show is called and media control is enabled") {
                        container.mediaControlEnabled = true
                        mediaControl.hide()
                        mediaControl.show()
                        
                        expect(mediaControl.controlsOverlayView.alpha) == 1
                        expect(mediaControl.controlsWrapperView.alpha) == 1
                        expect(mediaControl.controlsHidden).to(beFalse())
                    }
                }
                
                context("Play") {
                    it("Should call container play when is paused") {
                        mediaControl.playbackControlState = .Paused
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(container.isPlaying).to(beTrue())
                    }
                    
                    it("Should call container play when is stopped") {
                        mediaControl.playbackControlState = .Stopped
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(container.isPlaying).to(beTrue())
                    }
                    
                    it("Should trigger playing event ") {
                        var callbackWasCalled = false
                        mediaControl.once(MediaControlEvent.Playing.rawValue) { _ in
                            callbackWasCalled = true
                        }
                        
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        
                        expect(callbackWasCalled).to(beTrue())
                    }
                }
                
                context("Pause") {
                    beforeEach() {
                        mediaControl.playbackControlState = .Playing
                        playback.playbackType = .VOD
                    }
                    
                    it("Should call container pause when is playing") {
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(container.isPlaying).to(beFalse())
                    }
                    
                    it("Should change playback control state to paused") {
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(mediaControl.playbackControlState) == PlaybackControlState.Paused
                    }
                    
                    it("Should trigger not playing event when selecting button") {
                        var callbackWasCalled = false
                        mediaControl.once(MediaControlEvent.NotPlaying.rawValue) { _ in
                            callbackWasCalled = true
                        }
                        
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        
                        expect(callbackWasCalled).to(beTrue())
                    }
                }
                
                context("Stop") {
                    beforeEach() {
                        mediaControl.playbackControlState = .Playing
                        playback.playbackType = .Live
                        container.trigger(ContainerEvent.Ready.rawValue)
                    }
                    
                    it("Should call container pause when is live video is playing") {
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(container.isPlaying).to(beFalse())
                    }
                    
                    it("Should change playback control state to stopped") {
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(mediaControl.playbackControlState) == PlaybackControlState.Stopped
                    }
                    
                    it("Should trigger not playing event when selecting button") {
                        var callbackWasCalled = false
                        mediaControl.once(MediaControlEvent.NotPlaying.rawValue) { _ in
                            callbackWasCalled = true
                        }
                        
                        mediaControl.playbackControlButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        
                        expect(callbackWasCalled).to(beTrue())
                    }
                }
                
                context("Current Time") {
                    it("Should start with 00:00 as current time") {
                        expect(mediaControl.currentTimeLabel.text) == "00:00"
                    }
                    
                    it ("Should listen to current time updates") {
                        let info: EventUserInfo = ["position" : 78]
                        playback.trigger(PlaybackEvent.TimeUpdated.rawValue, userInfo: info)
                        
                        expect(mediaControl.currentTimeLabel.text) == "01:18"
                    }
                }
                
                context("Duration") {
                    it("Should start with 00:00 as duration") {
                        expect(mediaControl.currentTimeLabel.text) == "00:00"
                    }
                    
                    it ("Should listen to Ready event ") {
                        playback.trigger(PlaybackEvent.Ready.rawValue)
                        
                        expect(mediaControl.durationLabel.text) == "00:30"
                    }
                }
                
                context("End") {
                    it("Should reset play button state after container end event") {
                        mediaControl.playbackControlState = .Playing
                        container.trigger(ContainerEvent.Ended.rawValue)
                        
                        expect(mediaControl.playbackControlState) == PlaybackControlState.Stopped
                    }
                }
            }
        }
    }
    
    class StubedPlayback: Playback {
        var playing = false
        var playbackType = PlaybackType.VOD
        
        override func isPlaying() -> Bool {
            return playing
        }
        
        override func play() {
            playing = true
        }
        
        override func pause() {
            playing = false
        }
        
        override func type() -> PlaybackType {
            return playbackType
        }
        
        override func duration() -> Double {
            return 30
        }
    }
}